Fix issues working with large float values.

markstory opened the pull request:
Large float values were being truncated to scientific notation when being passed into PDO. This was caused by PHP's implicit float->string behavior that loses precision. Formatting into %F will retain all the necessary precision.

Refs #9424

markstory committed:
Fix issues working with large float values.

Large float values were being truncated to scientific notation when
being passed into PDO. This was caused by PHP's implicit float->string
behavior that loses precision. Formatting into %F will retain all the
necessary precision.

Refs #9424

markstory committed:
Fix incorrect type.

Column is an integer not a float.

chinpei215 commented on src/Database/Type/FloatType.php:
```diff
@@ -65,18 +65,18 @@ public function __construct($name = null)
         }
 
-        return (float)$value;
+        return sprintf('%F', (float)$value);
```
@markstory I think that we need to create DecimalType instead of mapping to FloatType. Strictly, decimal values cannot be mapped to float values. So `id` is declared as `number` instead of `float` in the table.
Reading #9424, the `id` is declared as number(**20**, 0), but float types cannot keep such large precision. This code would work when id is 123456789012345, but wouldn't work when id is 12345678901234567.

markstory replied:
In #9424 the [reflected schema](https://github.com/cakephp/cakephp/issues/9424#issuecomment-245237277) data has interpreted the column type as `decimal`.

Should we be truncating data that is going into decimal columns with lengths that are shorter than the data length?

chinpei215 replied:
First of all, I don't oppose this change. It could be better than now. Strictly, in MySQL, such big decimal values cannot be passed to the database as strings, [until MySQL 5.7.8](https://bugs.mysql.com/bug.php?id=72056). But now it works.

But, in my opinion, `decimal` shouldn't be truncated in any case. (In SQL server `numeric` is functionally equivalent to `decimal`. And, in my above comment, `number` is a typo. Sorry)

When people use `decimal` instead of `float`, they would need to store important data (e.g. money) to the field. They want to avoid to lose precision of the value. However, if we map `deciaml` to `float`, it would disappoint their expectation.

Now I think `decimal` should be mapped to `string` instead of `float`. It should be done in 3.4, as it would break BC. `DecimalType` class might be unnecessary. At first I thought we should create the class. But on second thoughts, there is not much we can do in the class in PHP.

markstory replied:
Wouldn't using the `StringType`  not fix the reported issue for native `float` values?

```php
php > $v = 123456789012345.0;
php > echo (string)$v;
1.2345678901234E+14
```

While this is what people are getting today because of how PDO works, I don't think it is right.

chinpei215 replied:
The exponential notation would work if the field is declared as `float` not `decimal`. The root cause of this issue is that mapping `decimal` to `FloatType`. If we map `decimal` to `StringType`, `$v` in your example will be a `string` instead of `float`. So the reported issue wouldn't happen.

Using `sprintf()` might fix #9424, if the reporter doesn't store number more than 16 digits into the field. But wouldn't it introduce another issue, in some edge case?
```php
$v = 9.9999999;
echo sprintf('%F', $v);
// 10.000000
```

markstory replied:
Using a `StringType` will result in incorrect types when data is pushed into entities. Previously entities would have floats, if we used `StringType` they would have strings which could cause problems with application math. 

We could add a new DecimalType that adds the proposed casting and leave the FloatType alone. Having a separate type would allow us to only format non-string values with sprintf().

chinpei215 replied:
OK. Now I think it would be better to keep as it is, at least regarding FloatType class.

What about adding a note to [the Data Types section of 3.x Cookbook](http://book.cakephp.org/3.0/en/orm/database-basics.html#data-types)? As of 3.x, this kind of issue could happen not only float, but also integer. `int unsigned` doesn't work as expected, if a number is greater than 2147483647 on 32-bit php. `bigint unsigned` never works on any platforms. If we want to use such data types, we need to map the field to `string` always. I think using `numeric(20, 0)` is the same situation.

inoas replied:
Decimal is expected to work pretty much like Integer and both are expected to never be automatically transformed, such as truncated. Not having worked with neither BC Math or GMP is what Int and Decimal may require?
http://stackoverflow.com/questions/211345/working-with-large-numbers-in-php http://www.php.net/manual/en/book.bc.php
http://www.php.net/manual/en/book.gmp.php

lorenzo replied:
I think that's an OK tradeoff @chinpei215. Big numbers are not supposed to work magically in 32-bit machines.

markstory commented:
Does this summary sound correct:

* Don't modify the existing `FloatType`
* Create a new `DecimalType` that:
  * Uses `sprintf()` like the changes currently does for floats, but leaves string data alone.
  * Continues to make floats as the PHP representation (as to not break existing apps)

lorenzo commented:
@markstory that sounds right to me

chinpei215 commented:
How does `DecimalType` know precision of decimal? If a field is declared as `decimal(M, 7)`, we need to pass `%.7F` to `sprintf()`.

markstory committed:
Create a standalone DecimalType.

This type replicates most of the behavior in the FloatType but includes
special affordances for numeric strings and ensures that precision is
not lost as frequently when data is bound to PDO.

Refs #9424

markstory commented:
DecimalType added now.

markstory commented:
@chinpei215 I'm not sure we can keep the precision from the column schema.

ADmad commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
```
3.3.4

markstory replied:
Fixed.

chinpei215 commented:
@markstory So if the decimal value is 0.9999999, it will be rounded up to 1.000000 unexpectedly. If users are using FloatType, it wouldn't lose precision, in contrast.

markstory commented:
@chinpei215 We could format to 8 decimal places by default. People needing more values could subclass/replace the decimal type?

markstory committed:
Fix doc tag.

markstory committed:
Fix failing test.

Marshalling oversize decimal values will result in float casting which
can lose precision. I'd like to remove that float casting but am
concerned that it will create backwards compatibility issues as people
would start getting strings instead of float values.

chinpei215 commented:
@markstory If we use `%.8F` by default, `70000000.1` would lose the precision:
```php
printf('%.8F', 70000000.1);
// 70000000.09999999
```

So I said `decimal` should be mapped to `string`. It should be done in 3.4, not in a bugfix release.

This change might allow to store 16-digit numbers into database without losing precision. But if people are satisfied with 16-digit numbers, I cannot understand why they choose decimal types. People can store them as `bigint` without losing precision. The maximum value of `bigint` is 19-digit. When I saw `numeric(20, 0)` in #9424, I thought the reporter really wanted to store 20-digit numbers.

markstory commented:
@chinpei215 I agree the PHP representation of decimals should be string for safer results.

markstory commented:
Is this OK to merge?

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Convert integer data into the database format.
+     *
+     * @param string|resource $value The value to convert.
```
Why would these be PHP resources?

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+            return null;
+        }
+        if (is_array($value)) {
+            return '1';
```
Why return '1' if submitted value is an array?

markstory replied:
Because arrays are invalid, and throwing an exception doesn't seem very helpful. This behavior is also consistent with the FloatType.

lorenzo replied:
Actually we do throw exceptions in `toDatabase`, it is a good idea to follow that https://github.com/cakephp/cakephp/blob/master/src/Database/Type/IntegerType.php#L44-L45

markstory replied:
Done.

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+            return $value;
+        }
+
+        return sprintf('%F', (float)$value);
```
Why the cast to float? Why not to (string)? What should this case catch anyway? PHP ressource type?

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    /**
+     * Convert float values to PHP integers
```
Shouldn't DecimalType::toPHP convert to string to have arbitrary precision?

markstory replied:
Ideally yet, but I have worries about breaking people's code.

inoas replied:
So because 'decimal' => 'Cake\Database\Type\DecimalType', pointed to float before this cannot/should not be fixed? Maybe we want something like BigNumType and have that behave with absolute precision and then map decimal there OPTIONALLY? Maybe it is a thing to be done for 4.0? However the docs should really have a fat warning that decimal type is not precise... I'd always expect decimal to behave like integer (or string) and never like float.

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+            return null;
+        }
+        if (is_array($value)) {
+            return 1;
```
Why would the value coming from the database be of type array? Why should it then return int 1

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+            return 1;
+        }
+
+        return (float)$value;
```
See above

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    /**
+     * Marshalls request data into PHP floats.
```
Same as above (why not to string)?
Same questions within the marshal method, as before

inoas commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    /**
+     * Converts a string into a float point after parseing it using the locale
```
Again floats

inoas commented on tests/TestCase/Database/Type/DecimalTypeTest.php:
```diff
@@ -0,0 +1,178 @@
+    {
+        $this->assertNull($this->type->toPHP(null, $this->driver));
+
+        $result = $this->type->toPHP('some data', $this->driver);
```
Shouldn't this throw an exception instead?

lorenzo replied:
This is data coming from the database. It would be annoying having to deal with a conversion error at this level.

inoas commented on tests/TestCase/Database/Type/DecimalTypeTest.php:
```diff
@@ -0,0 +1,178 @@
+        $this->assertSame(2.0, $result);
+
+        $result = $this->type->toPHP('2 bears', $this->driver);
+        $this->assertSame(2.0, $result);
```
Shouldn't this throw an exception instead?

inoas replied:
Sure but this is an abstraction on top of PHP that can be more strict than vanilla PHP. It is a Type class after all?

inoas commented on tests/TestCase/Database/Type/DecimalTypeTest.php:
```diff
@@ -0,0 +1,178 @@
+        $this->assertSame(2.51, $result);
+
+        $result = $this->type->marshal('3.5 bears', $this->driver);
+        $this->assertSame('3.5 bears', $result);
```
Shouldn't this throw an exception instead?

inoas replied:
Good point!

markstory committed:
Make return data more consistent.

Return floats where possible, and remove unnecessary code from toPHP()
as the database will not return arrays from decimal columns.

lorenzo commented on tests/TestCase/Database/Type/DecimalTypeTest.php:
```diff
@@ -0,0 +1,178 @@
+    public function testMarshal()
+    {
+        $result = $this->type->marshal('some data', $this->driver);
+        $this->assertSame('some data', $result);
```
No, marshal should never throw an exception

inoas commented:
I am not certain this is a good idea. Switching from floats to integer (with exceptions being thrown on breaking boundaries) or to strings makes a lot of sense. Adding a DecimalType that behaves like floats not like integers/bignum/fixnum feels wrong to me. If this goes through I'd love to see a big red warning in the docs and hopefully have it fixed (non BC) for 4.0.

dereuromark commented on src/Database/Type/DecimalType.php:
```diff
@@ -0,0 +1,177 @@
+
+            return $this;
+        }
+        if (static::$numberClass === 'Cake\I18n\Number' ||
```
Could be use ::class syntax here as this is available since 5.5+ afaik.

markstory commented:
>  Adding a DecimalType that behaves like floats not like integers/bignum/fixnum feels wrong to me. If possible I'd then still link it to FloatType - at least then you get what you expect.

We need to both preserve the existing behavior and also fix the issues around very large numbers. By adding a new Type I can not impact FloatType at all, fix the issues with decimals losing precision.

jimgwhit commented:
In php / mysql remember these 3 rules  

* Do not use floats
* Do not use floats
* Do not use floats  

Use decimal

ADmad commented:
@jimgwhit Please avoid posting non-constructive comments on issues.

markstory commented:
Doing math in PHP space without floats is complicated as it requires extensions like gmp or bcmath.

jimgwhit commented:
@ADmad I have been dealing with this stuff a long time, floats can be a freaking mess, I meant it about use decimal, or anything except float. They can be a pain.

markstory committed:
Throw exceptions when decimal type gets a non-scalar.

We can't convert non-scalar values into floats for use in database
statements.

markstory commented:
Is this ok now @lorenzo ?

lorenzo commented:
Looks good :)
